server/src/main/java/com/google/sps/utils/StateSubcollectionBuilder.java

```diff
@@ -0,0 +1,140 @@
+ *              each advertiser's total ad spend, and 'stateData', which contains
+ *              information about each state's total ad spend.
+ * Author: Kira Toal
+ * Date: July 20, 2020
+ */ 
+public class StateSubcollectionBuilder {
+
+  private static final String PATH_TO_SERVICE_ACCOUNT = "./serviceAccountKey.json"; 
+  private static final String DATABASE_URL = "https://step9-2020-capstone.firebaseio.com"; 
+  private static final String MAIN_COLLECTION = "ads";
+  private static final String WRITE_COLLECTION = "dev_states";
+  private static final Set<String> GEO_TARGETS = Constants.TEST_GEO_TARGETS;
+  public static Firestore db; 
+
+  public static Firestore initializeApp() throws Exception {
+    // Set account and build options.
+    FileInputStream serviceAccount = new FileInputStream(PATH_TO_SERVICE_ACCOUNT);
+    FirebaseOptions options = new FirebaseOptions.Builder()
+        .setCredentials(GoogleCredentials.fromStream(serviceAccount))
+        .setDatabaseUrl(DATABASE_URL)
+        .build();
+    
+    // Initialize app.
+    if (FirebaseApp.getApps().isEmpty()) {
+      FirebaseApp.initializeApp(options);
+    }
+    db = FirestoreClient.getFirestore();
+    return db;
+  }
+
+  /**
+   * Maps advertisers to their total ad spend in a particular state
+   * @param documentsInState list of all document snapshots targeted
+   *                         at a particular state 
+   */
+    public static Map<String, Long> getAdvertiserToSpend(
+      List<QueryDocumentSnapshot> documentsInState) {
+    Map<String, Long> advertiserToSpend = new HashMap<>();
+
+    for (QueryDocumentSnapshot document: documentsInState) {
+      String advertiser = document.getString("advertiser");
+      long spendMax = document.getLong("spendMax");
+
+      if (advertiserToSpend.containsKey(advertiser)) {
+        spendMax += advertiserToSpend.get(advertiser); 
+      }
+      
+      advertiserToSpend.put(advertiser, spendMax);
+    }
+    
+    return advertiserToSpend;
+  }
+
+  /**
+   * Writes advertiser documents to Firestore
+   * @param state the current state subcollection (tells Firestore
+   *              where to write the advertiser documents)
+   * @param advertiserToSpend map that relates every advertiser
+   *              to their total ad spend in a particular state              
+   */
+  public static void updateStateSubcollection(String state,
+      Map<String, Long> advertiserToSpend) {
+
+    long totalStateSpend = 0;
+
+    for (String key : advertiserToSpend.keySet()) {
+      Map<String, Object> data = new HashMap<>();
+      long totalAdvertiserSpend = advertiserToSpend.get(key);
+      totalStateSpend += totalAdvertiserSpend;
+      data.put("totalAdvertiserSpend", totalAdvertiserSpend);
+      db.collection(WRITE_COLLECTION)
+        .document(state.toLowerCase())
+        .collection("advertisers")
+        .document(key)
+        .set(data, SetOptions.merge()); // Granular merge instead of overwrite.
```

Comment:
Does Firestore have some kind of bulk update API, so you don't have to do this one-by-one in a loop?